pkg/query/endpointset.go

```diff
@@ -277,7 +279,12 @@ func (c *endpointSetNodeCollector) Collect(ch chan<- prometheus.Metric) {
 					lbls = append(lbls, storeTypeStr)
 				}
 			}
-			ch <- prometheus.MustNewConstMetric(c.connectionsDesc, prometheus.GaugeValue, float64(occurrences), lbls...)
+			select {
+			case ch <- prometheus.MustNewConstMetric(c.connectionsDesc, prometheus.GaugeValue, float64(occurrences), lbls...):
```

Comment:
I think something else might be at play here, e.g., this goroutine is starved of CPU resources. Do you have a lot of goroutines running when this happens?

Reply:
There is only 1 goroutine here but there are new goroutines constantly piling up [here](https://github.com/thanos-io/thanos/blob/main/pkg/query/endpointset.go#L516). That is because the mutex for the collector is locked, which causes the call [here](https://github.com/thanos-io/thanos/blob/main/pkg/query/endpointset.go#L459) to hang so [this](https://github.com/thanos-io/thanos/blob/main/pkg/query/endpointset.go#L426) mutex also is locked.